diff --git a/services/credentials/java/com/android/server/credentials/CredentialManagerService.java b/services/credentials/java/com/android/server/credentials/CredentialManagerService.java
index bc5c427e3ccb..9960fe8b03fa 100644
--- a/services/credentials/java/com/android/server/credentials/CredentialManagerService.java
+++ b/services/credentials/java/com/android/server/credentials/CredentialManagerService.java
@@ -1236,29 +1236,29 @@ public final class CredentialManagerService
 
             // Set the filtered primary providers to the settings key
             if (!settingsWrapper.putStringForUser(
                     Settings.Secure.CREDENTIAL_SERVICE_PRIMARY,
                     String.join(":", filteredPrimaryProviders),
-                    UserHandle.myUserId(),
+                    userId,
                     /* overrideableByRestore= */ true)) {
                 Slog.e(TAG, "Failed to remove primary service: " + componentName);
                 return;
             }
         }
 
         // Read the credential providers to remove any reference of the removed service.
         String rawCredentialProviders =
                 settingsWrapper.getStringForUser(
-                        Settings.Secure.CREDENTIAL_SERVICE, UserHandle.myUserId());
+                        Settings.Secure.CREDENTIAL_SERVICE, userId);
 
         // Remove any provider services that are same as the one being removed.
         Set<String> filteredCredentialProviders = getStoredProvidersExceptService(
                 rawCredentialProviders, componentName);
         if (!settingsWrapper.putStringForUser(
                 Settings.Secure.CREDENTIAL_SERVICE,
                 String.join(":", filteredCredentialProviders),
-                UserHandle.myUserId(),
+                userId,
                 /* overrideableByRestore= */ true)) {
             Slog.e(TAG, "Failed to remove secondary service: " + componentName);
         }
     }
 
